### PR TITLE
refactor: rewrite lsp to use lspower

### PR DESCRIPTION
**Pull request by lucacasonato**
This PR investigates if we could use `tower-lsp` as our LSP server framework.

`tower-lsp` is an async LSP implementation written using `tower` (same framework `hyper` and `tonic` are built on). It uses the same underlying `lsp_types` crate for the protocol as `lsp_server`.

It has some nice usability benefits over `lsp_server`:
- It is async out of the box.
- It has built in notification and request routing.
- It has a less verbose API than `lsp_server`. 

TODO:
- [x] port over TSC and lint integration
- [x] port over tests
- [x] wait for https://github.com/ebkalderon/tower-lsp/pull/243 to land and be released, or publish to `tower-lsp`.
- [x] remove unused code

### CODE REVIEW

File: cli/Cargo.toml
Comment (kitsonk) [inline cli/Cargo.toml:83]: Probably should consider waiting on this until these become true, IMO.

**Review by kitsonk — COMMENTED** [review]
Overall, it feels like the right track but a couple of things:

- renaming `lsp` to `lsp2` just makes the PR infinitely more difficult to read and feels totally arbitrary.  If we had to have two language servers running for a while, then we would want to use `lsp2` but this PR fully removes `lsp` in lieu of `lsp2`.
- some of the arbitrary renaming of things also makes it difficult to read, there is no problem with doing a rename PR as a seperate PR first.
- I really think we should hold off on it until we a stable crate published to replace lsp server.  There is no harm in waiting a little while to have the PRs accepted or the alternative crate published.

I know we landed the original LSP as a big ole PR, but it was hard to avoid that... I think we should really try to do things in smaller chunks now.

**Comment by lucacasonato** [comment]
> renaming lsp to lsp2 just makes the PR infinitely more difficult to read and feels totally arbitrary. If we had to have two language servers running for a while, then we would want to use lsp2 but this PR fully removes lsp in lieu of lsp2.

Yes - I am going to remove lsp 1 in and move to rename lsp2 to lsp. It was just easier to copy paste chunks in small batches rather than porting everything in the same commit :-)

> Some of the arbitrary renaming of things also makes it difficult to read, there is no problem with doing a rename PR as a seperate PR first.

I dont think I did any major renames. Could you point me to what you mean?

> I really think we should hold off on it until we a stable crate published to replace lsp server. There is no harm in waiting a little while to have the PRs accepted or the alternative crate published.

Yup - but mainly a chicken and egg problem. I can get this out to crates.io, but I am not going to unless we agree that we want to go with `lsp-server`. Don't want to push a new crate to cargo, just to determine we won't actually use it.

> I know we landed the original LSP as a big ole PR, but it was hard to avoid that... I think we should really try to do things in smaller chunks now.

Agree - but I don't think that is possible with this PR. It completely replaces the core of the lsp. If you have ideas about how we could do this, I can split it up.

File: cli/Cargo.toml
Comment (lucacasonato) [reply cli/Cargo.toml:83]: Yes this is a merge blocker.

**Comment by kitsonk** [comment]
> I dont think I did any major renames. Could you point me to what you mean?

```diff
- server_state
+ state_snapshot
- ServerStateSnapshot
+ StateSnapshot
```

I'm fine with the change, but there were a lot of lines in this PR that were just "noise" around those renames.

> Yup - but mainly a chicken and egg problem.

How big would `lspower` be?  How much of `tower-lsp` would we not be using?  Does the server need to keep current with tokio?  In my mind if it is small, there are parts of `tower-lsp` we would never use, and/or we have to keep it current with other parts of Deno like tokio, then we should consider just having our own thing (or building the server directly into the lsp).

**Comment by lucacasonato** [comment]
> How much of lspower would we not be using?

We are pretty much using all of lspower. We dont support all language server APIs, but there is no big feature we dont use.

> Does the server need to keep current with tokio?

It needs to update to 0.3 when we do. I checked, and this is trivial. Only 3 lines change compared to 0.2.

File: cli/lsp/README.md
Comment (ry) [inline cli/lsp/README.md:23]: unintentional?

Comment (lucacasonato) [reply cli/lsp/README.md:23]: Yeah. Whoops

File: cli/lsp/backend.rs
Comment (kitsonk) [inline cli/lsp/backend.rs:41]: nit: I really hate the term "backend"... `LanguageServer`?

Comment (kitsonk) [reply cli/lsp/backend.rs:41]: or `DenoLanguageServer` so as to not conflict with the trait?

File: cli/lsp/lsp_extensions.rs
Comment (kitsonk) [inline cli/lsp/lsp_extensions.rs:10]: probably should be moved to "backend" (which I still don't like the name) and this file deleted.

File: cli/lsp/utils.rs
Comment (kitsonk) [inline cli/lsp/utils.rs:8]: Feels like this could be moved somewhere now, as we don't really have much here...

Comment (kitsonk) [reply cli/lsp/utils.rs:8]: actually they are duplicated in `cli/ops/util.rs` now too, though they are only specific to lsp ops at the moment

**Review by kitsonk — COMMENTED** [review]
Looking good... the term "backend" is really bothering me though.

File: cli/lsp/backend.rs
Comment (lucacasonato) [reply cli/lsp/backend.rs]: Renamed to `DenoLanguageServer`

**Review by kitsonk — APPROVED** [review]
❤️ 

I guess it is just a decision about what we do about waiting for lspower.  I will defer to @ry on that one.

File: cli/lsp/backend.rs
Comment (ry) [inline cli/lsp/backend.rs:41]: How about just `LanguageServer`?

Comment (kitsonk) [reply cli/lsp/backend.rs:41]: and then use `lspower::LanguageServer` for the trait?

Comment (lucacasonato) [reply cli/lsp/backend.rs:41]: SGTM

**Comment by lucacasonato** [comment]
lspower is published to crates now. Just need to switch the cargo toml line.￼

Also still need to fix the unit test in backend.rs and rename that file to language_server.rs.

**Pull request merged**
